pkg/gcmanager/gc.go

```diff
@@ -0,0 +1,556 @@
+		go s.tracePodWorker(ctx)
+	}
+
+	for i := 0; i < RELEASE_IPPOOLIP_WORKER_NUM; i++ {
+		go s.releaseIPPoolIPWorker(ctx)
+	}
+
+	for i := 0; i < RELEASE_WEP_WORKER_NUM; i++ {
+		go s.releaseWEPWorker(ctx)
+	}
+
+	go s.monitorCLISignal(ctx)
+	go s.monitorDefaultGCInterval(ctx)
+
+	logger.Info("Running gc...")
+}
+
+func (s *SpiderGC) GetPodDatabase() {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (s *SpiderGC) TriggerGCAll() {
+	s.gcCliSignal <- struct{}{}
+}
+
+func (s *SpiderGC) Health() {
+	//TODO implement me
+	panic("implement me")
+}
+
+// ----------------------------------------------------------------------------------------------------
+
+// registerPodReconcile registers watch pod
+func (s *SpiderGC) registerPodReconcile() error {
+	c, err := controller.New("watch-pod-controller", s.ControllerManager, controller.Options{
+		Reconciler: &reconcilePod{
+			spiderGC: s,
+		},
+	})
+	if nil != err {
+		return err
+	}
+
+	// TODO 后面的predicate用于过滤，是否需要Genric事件？
+	err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForObject{})
+	if nil != err {
+		return err
+	}
+
+	return nil
+}
+
+// executeScanAllPod scans the whole pod and whole IPPoolList
+func (s *SpiderGC) executeScanAllPod(ctx context.Context) {
+	// 查出来的
+	podList := &corev1.PodList{}
+	poolList := &v1.IPPoolList{}
+
+	// 场景1：k8s中不存在的pod对象，但是存在于ippool里的ip-allocationDetail里
+	var foundInPod string
+	for _, pool := range poolList.Items {
+		for ippoolIP, poolIPAllocation := range pool.Status.AllocatedIPs {
+			for _, pod := range podList.Items {
+
+				// ippool记录中的pod在k8s中有
+				if poolIPAllocation.Pod == pod.Name && poolIPAllocation.Namespace == pod.Namespace {
+					foundInPod = pod.Name
+
+					s.filterIP(ctx, &pod, pool, ippoolIP)
+					break
+				}
+			}
+
+			// 场景1：k8s中不存在的pod对象，但是存在于ippool里的ip-allocationDetail里
+			if len(foundInPod) == 0 {
+				s.gcIPPoolIPSignal <- gcIPPoolIPIdentify{
+					Pool:         &pool,
+					IP:           ippoolIP,
+					PodName:      poolIPAllocation.Pod,
+					PodNamespace: poolIPAllocation.Namespace,
+					IsReleaseWEP: true,
+				}
+				continue
+			}
+
+			// 找到后就置空，等下一个ippoolIP
+			foundInPod = ""
+		}
+	}
+
+	return
+}
+
+// filterIP check the given pod whether matches the ippoolIP, if not then gc the bad ippoolIP
+func (s *SpiderGC) filterIP(ctx context.Context, pod *corev1.Pod, pool v1.IPPool, ippoolIP string) {
+	// 场景2：防止watch遗漏? (ippool记录中的pod在k8s中有)：k8s中pod存在，pod属于terminating状态,
+	// 如果cache里面有记录且状态已经标志为terminating则表明已经被追踪，否则就更新cache且发信号
+	if nil != pod.DeletionTimestamp {
+		podCache, err := s.PodDB.Get(pod.Name, pod.Namespace)
+		if nil != err {
+			// 1.有可能已经被gc  2.有可能watch丢失. 如果是已经被gc，没关系重做一次，后续逻辑都能正常。
+			podEntry, err := s.PodDB.Create(pod)
+			if nil != err {
+				// 实际上此处就不可能出现err的，哪怕有也无所谓。等下一轮gcAll继续处理
+				logger.Sugar().Errorf("error: create pod '%+v' cache  failed '%v', concurrency conflicts", types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}, err)
```

Comment:
logger.Sugar().Errorf("scan %v in ippool %v, failed to create cache for a deleting pod %v/%v , reason=%v \n", ippoolIP , poolName,  pod.Namespace, pod.Name, err)